### DISCUSSION TITLE
Multiple forms for a single module

### DISCUSSION THREAD

**Discussion by wratke** [area17/twill#907] [Q&A] [answered]
I'm trying to think of the best way to add pages (similar to the Pages in Statamic: https://docs.statamic.com/pages). It seems the best way to do this is create a ```pages``` module, however, each page would need a custom form associated with it. Is this possible?

**Comment by ifox** [comment, marked as answer]
Hi @wratke!

I see 3 possibles approaches there, depending on the number of pages you'll need and their forms compositions:

1. A single `pages` module, with all fields and relationships in the same Eloquent model. This can seem ugly, but the simplest way to implement if you don't have tons of pages with very different fields. In the form's Blade file, you can include a different file depending on a type you store in pages records (seeded upfront or created through the UI, with a dropdown to chose type inside the Add new modal, which you can add by overwriting `create.blade.php` in your module views).

2. One module per page, which means a single record in each associated table, and no listing in the CMS, but direct access to the form through the navigation, by using a route name in `twill-navigation.php` instead of `'module' => true`. This is working great for say big landing pages that requires a lot of copy fields, browsers, repeaters and so on. Approach 1 becomes quickly messy in those instances.

3. A single `pages` module, with all fields saved in a JSON column. This would require some custom logic in your module's repository, but this is definitely possible. My comment about page's type in 1. also apply here.

Most of the time for us at AREA 17, projects use a combination of 1 and 2, as well as a "generic" page model where all records share the same form but users are benefiting from the block editor to create any composition on those.

**Comment by wratke** [comment]
Thanks, this is helpful. Is it possible to use a route that points directly to a specific record in the database in the ```twill-navigation.php``` file? Say I create a module for a page and created a single record (which had an id of 1). Could I create a route that links directly to the edit page for that specific record using it's id?

**Comment by ifox** [comment]
You could do it using the id, yes, by using the `'raw'` option in a navigation entry:

```php
'page' => [
    'raw' => true,
    'route' => /pages/1/edit
]
```

However, you can also register a custom route for it:

```php
Route::module('pages');
Route::name('homepage')->get('homepage', 'PageController@homepage');
```

Then, in the `pages` module controller:

```php
public function homepage()
{
    // find the page by type or anything you've decided to use to differentiate them
    abort_unless($page = Page::whereType('homepage')->first(), 500, 'Homepage missing');
    
    // this is so that Twill always goes back to that form's route directly, not pages listing
    Session::put("pages_back_link", route('admin.homepage'));
    
    // with this approach you can provide your own view for each page
    return view('admin.pages.homepage',  $this->form($page->id));
}
```

**Comment by wratke** [comment]
Great, thanks!

**Comment by wratke** [comment]
One quick question. Is it possible to access the data that populates the form in the blade template? Is that data accessible as a PHP variable ? I can see how to create conditional logic in the blade template in the Laravel docs, but I need to get a sense of how Twill handles the data behind the scenes before I can do that.

**Comment by wratke** [comment]
I'm talking specifically about the form.blade.php template generated for a module. If I have a field in the database called ```type```, I'm trying to figure out how to access that data from whatever Twill is doing when it populates the form in the template.

**Comment by ifox** [comment]
You should be able to access the edited model using the `$item` variable in Blade.

**Comment by wratke** [comment]
It works! Thanks.

**Comment by istvanvasil** [comment]
> 2. One module per page, which means a single record in each associated table, and no listing in the CMS, but direct access to the form through the navigation, by using a route name in `twill-navigation.php` instead of `'module' => true`. This is working great for say big landing pages that requires a lot of copy fields, browsers, repeaters and so on. Approach 1 becomes quickly messy in those instances.

Hi @ifox 

I've been trying to wrap my head around how you create single pages in Twill. The above method sounds like the most logical in many cases when you want to keep your pages as separate from each other as possible. I just have a hard time figuring out what is the so to speak 'best practice' to go about it. I am new to Laravel so I tried to avoid writing obvious questions but I am slightly stuck on this one.

I basically need to create a simple About page and the way I did it is that I created a CRUD module `twill:module -BSM abouts` then in the `twill-navigation.php` i did the following:

`'abouts' => [
	'title' => 'About',
	'route' => 'admin.abouts.show',
	'params' => [ 'about' => 1 ],
]`

It definitely solves my problem but the question is is this the "right" way to do it or will it cause problems down the line? Is there a more elegant way to do this? My frontend will be consumed through graphql and completely detached from the twill/laravel part of the site.

Thank you!

**Comment by devingray** [comment]
I managed to get this to work really nicely with a custom approach. 

Firstly my migration file will have these fields

```
// Stores a path to a form blade file
$table->string('form')->nullable();
// Stores a path to a view blade file
$table->string('view')->default('front.pages.default');
// Store data
$table->json('data')->nullable();
```
When a page is created I have an observer that runs the following:
```
function copyFormAndViewStubs($name): void
    {
        $formPath = resource_path("views/admin/pages/_${name}.blade.php");
        $viewPath = resource_path("views/front/pages/${name}.blade.php");
        if (! File::exists($formPath)) {
            File::copy(
                resource_path('stubs/admin/_form.blade.php.stub'),
                $formPath
            );
        }
        if (! File::exists($viewPath)) {
            File::copy(
                resource_path('stubs/pages/front/view.blade.php.stub'),
                $viewPath
            );
        }
    }
```

This copies a simple stub from a stubs folder into two places

`admin/pages/_name.blade.php` and `front/pages/name.php`

Then my frontend controller and routes to display pages looks like this 

```
// Route
Route::get('/{slug?}', PageController::class)->name('page.show');

// Method
public function __invoke(PageRepository $repository, ?string $slug = null): PageResponse
    {
        abort_unless($item = isset($slug)
            ? $repository->forSlug($slug)
            : $repository->forSlug('home'), 404);

        return view($this->item->view, [
            'item' => $item,
            'data' => json_decode($item->data, true),
        ]);
    }
```

in my cms form I simply include the form blade file

```
@section('contentFields')
    @formField('input', [
        'name' => 'title',
        'label' => 'Title',
        'required' => true,
    ])
    @include($item->form)
@endsection
```

Now can build custom forms for each page so long as I specify which fields are savable in the json field
```
protected $fieldsGroups = [
        'data' => [
            'example',
        ]
    ];
```

**Reply by ifox** [reply]
Such a simple yet empowering way of using Twill features @devingray, thanks for sharing.
